k2/addons/k2-serial/zserial/serial.py

```diff
@@ -54,6 +55,75 @@ def serial_exec(sut, messagebus):
     return SerialClient(messagebus, sut.entity, sut.serial.timeout, sut.serial.prompt)
 
 
+@requires(sut='Sut', can=['serial'])
+@requires(messagebus='MessageBus')
+@component(name='RawSerialPort', can=['serial'], provided_by_extension='zserial')
+class RawSerialPort(object):
+    """
+    Access the serial port in raw byte mode.
+
+    Temporarily suspend the packetized log line handling to provide raw byte
+    access to the serial port.
+
+    Example:
+
+    .. code-block:: python
+
+        @requires(serial='RawSerialPort')
+        @requires(exec='Exec', can=['serial'])
+        def test_xmodem_recv(serial,exec):
+            with open('/dev/null', 'wb') as f:
+                exec.send_line('sz -X --1k -q /proc/cpuinfo',
+                    prefix_output=False,
+                    endmark='Give your local XMODEM receive command now.')
+                with serial.suspended() as ser:
+                    def getc(size, timeout=1):
+                        return ser.read(size) or None
+
+                    def putc(data, timeout=1):
+                        return ser.write(data)
+
+                    modem = XMODEM1k(getc, putc)
+                    modem.recv(f)
+
+        @requires(serial='RawSerialPort')
+        @requires(exec='Exec', can=['serial'])
+        def test_xmodem_send(serial,exec):
+            with open('/proc/version', 'rb') as f:
+                exec.send_line('rz -X -q -y /dev/null',
+                    prefix_output=False,
+                    endmark='rz: ready to receive /tmp/foo')
+                with serial.suspended() as ser:
+                    def getc(size, timeout=1):
+                        return ser.read(size) or None
+
+                    def putc(data, timeout=1):
+                        return ser.write(data)
+
+                    modem = XMODEM1k(getc, putc)
+                    modem.send(f)
+    """
+
+    def __init__(self, sut, messagebus):
+        self._sut = sut
+        self._messagebus = messagebus
+
+    @contextmanager
+    def suspended(self):
+        """Suspend the normal serial log line handling."""
+
+        try:
+            logger.debug(
+                'Suspending serial port for entity {entity}'.format(entity=self._sut.entity))
+            port = self._messagebus.send_request(
+                SERIAL_SUSPEND, SERIAL_ENDPOINT, entity=self._sut.entity).wait()[0].result()
+            yield port
+        finally:
+            logger.debug('Resuming serial port for entity {entity}'.format(entity=self._sut.entity))
+            self._messagebus.send_request(
+                SERIAL_RESUME, SERIAL_ENDPOINT, entity=self._sut.entity).wait()[0].result()
```

Comment:
The implementation can be extracted into two methods `.suspend()` and `.resume()`, called from the try- and finally-block.
That way the component can be used without using the context manager and without resorting to ugliness such as `.suspended().__enter__()`.

Reply:
Good suggestion! Fixed in latest diff :wink: